### PR TITLE
Feature/cppserver

### PR DESCRIPTION
**Pull request by tarc**
Specify library name and version:  **cppserver/1.0.0.0**

- [X] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [X] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [X] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [X] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by tarc** [comment]
The patches in this PR are:

- `00001-cmake-clean-up.patch`: to compatibilize main `CMakeLists.txt` to interact with Conan;
- `00002-define-win32-winnt.patch`: to fix issues that arise when building with mingw gcc;
- `00003-remove-asio-defines.patch`: to use Conan Center's ASIO standalone package which is header only, whereas upstream expects a compiled library;
- `00004-add-mbig-obj-to-mingw.patch`: again, to fix some issues found when building with mingw gcc.

File: recipes/cppserver/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    description = "Ultra fast and low latency asynchronous socket server and" \
+        " client C++ library with support TCP, SSL, UDP, HTTP, HTTPS, WebSocket" \
+        " protocols and 10K connections problem solution."
+    topics = ("network", "socket", "async", "low-latency")
+    settings = "os", "compiler", "build_type", "arch"
+    options = {"fPIC": [True, False],
+               "shared": [True, False]}
+    default_options = {"fPIC": True,
+                       "shared": False}
+    requires = ["asio/1.17.0", "openssl/1.1.1g", "cppcommon/cci.20201104"]
```
Comment (madebr) [inline recipes/cppserver/all/conanfile.py:20]:
```suggestion
    requires = ["asio/1.17.0", "openssl/1.1.1i", "cppcommon/cci.20201104"]
```

File: recipes/cppserver/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+            self._cmake = CMake(self)
+            self._cmake.definitions["CPPSERVER_MODULE"] = "OFF"
+            self._cmake.configure(build_folder=self._build_subfolder)
+        return self._cmake
+
+    def source(self):
+        tools.get(**self.conan_data["sources"][self.version])
+        extracted_dir = glob.glob("CppServer-*")[0]
+        import time
+        time.sleep(30)
```
Comment (madebr) [inline recipes/cppserver/all/conanfile.py:44]: Why was this necessary?
```suggestion
```

File: recipes/cppserver/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        if self.settings.os == "Windows":
+            del self.options.fPIC
+
+    def configure(self):
+        if self.options.shared:
+            del self.options.fPIC
+
+    def build(self):
+        patches = self.conan_data["patches"][self.version]
+        for patch in patches:
```
Comment (madebr) [inline recipes/cppserver/all/conanfile.py:57]:
```suggestion
        for patch in self.conan_data.get("patches", {}).get(self.version, []):
```

File: recipes/cppserver/all/patches/00001-cmake-clean-up.patch [hunk 4]
```diff
@@ -0,0 +1,141 @@
+diff --git a/CMakeLists.txt b/CMakeLists.txt
+index 5a21eed9..94135f07 100644
+--- a/CMakeLists.txt
++++ b/CMakeLists.txt
+@@ -18,50 +18,50 @@ if(DOXYGEN_FOUND)
+ endif()
+ 
+ # CMake module path
+-set(CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/cmake")
```
Comment (madebr) [inline recipes/cppserver/all/patches/00001-cmake-clean-up.patch:9]: My only comment about this patch is that it is hard to maintain for future cppserver releases.

Ideal is to create a patch adding options that can be set in the conanfile.
A patch could then be sent upstream.

File: recipes/cppserver/all/conanfile.py (hunk 2 above)
Comment (tarc) [reply recipes/cppserver/all/conanfile.py:44]: Oh, boy, sorry for that. It slipped in.

I have an ongoing issue with an antivirus here on my notebook which requires such kind of dirty tricks to bypass.

Comment (madebr) [reply recipes/cppserver/all/conanfile.py]: Don't worry. At one time, we have all met a hard to understand set-up.

Maybe you can create a little `post_source` hook with a sleep on your problematic setup?
https://docs.conan.io/en/latest/extending/hooks.html

File: recipes/cppserver/all/patches/00001-cmake-clean-up.patch (hunk 4 above)
Comment (tarc) [reply recipes/cppserver/all/patches/00001-cmake-clean-up.patch:9]: Yes, I agree. The other projects of the same author that are dependencies of this one suffer from the same issue.

Maybe we can go on with this PR the way it is for now, and I commit myself to send PRs upstream to handle this more smoothly.

What do you think?

File: recipes/cppserver/all/patches/00001-cmake-clean-up.patch [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+ 
+ # Library
+ file(GLOB_RECURSE LIB_HEADER_FILES "include/*.h")
+@@ -70,14 +70,15 @@ file(GLOB_RECURSE LIB_SOURCE_FILES "source/*.cpp")
+ add_library(cppserver ${LIB_HEADER_FILES} ${LIB_INLINE_FILES} ${LIB_SOURCE_FILES})
+ set_target_properties(cppserver PROPERTIES COMPILE_FLAGS "${PEDANTIC_COMPILE_FLAGS}" FOLDER "libraries")
+ target_include_directories(cppserver PUBLIC "${CMAKE_CURRENT_SOURCE_DIR}/include")
+-target_link_libraries(cppserver ${LINKLIBS} asio)
++target_link_libraries(cppserver ${LINKLIBS} CONAN_PKG::asio)
++target_compile_features(cppserver PUBLIC cxx_std_17)
```
Comment (prince-chrismc) [inline recipes/cppserver/all/patches/00001-cmake-clean-up.patch:87]: this seems out of place 🤔

Comment (tarc) [reply recipes/cppserver/all/patches/00001-cmake-clean-up.patch:87]: You mean the `cxx_std_17` feature, right?

This is similar to #4001.

This is one of those cases where the author handles this upstream with a CMake module. The relevant parts are:

- `CppServer/.gitlinks`:
```
# Scripts
build build https://github.com/chronoxor/CppBuildScripts.git master
cmake cmake https://github.com/chronoxor/CppCMakeScripts.git master
```
https://github.com/chronoxor/CppServer/blob/c67a656cac13c86e005f6ffcb288b9ffa8daf132/.gitlinks#L8-L10

- `CppServer/CMakeLists.txt`:
```cmake
# Compiler features
include(SetCompilerFeatures)
include(SetCompilerWarnings)
include(SetPlatformFeatures)
include(SystemInformation)
```
https://github.com/chronoxor/CppServer/blob/c67a656cac13c86e005f6ffcb288b9ffa8daf132/CMakeLists.txt#L23-L27

- `CppCMakeScripts/SetCompilerFeatures.cmake`:
```cmake
# Compiler features


# Choose C++ standard
set(CMAKE_CXX_STANDARD 17)
```

https://github.com/chronoxor/CppCMakeScripts/blob/4152297a1c28cfd2ea63ea9a4bdfcb50aa2c05d9/SetCompilerFeatures.cmake#L1-L4

In the other PR I said that this is done through git submodules, but I was mistaken. It's actually handled with a custom dependency management system (the author calls it GIL).

File: recipes/cppserver/all/patches/00001-cmake-clean-up.patch (hunk 4 above)
Comment (uilianries) [reply recipes/cppserver/all/patches/00001-cmake-clean-up.patch:9]: > Maybe we can go on with this PR the way it is for now, and I commit myself to send PRs upstream to handle this more smoothly.

Sounds good

Comment (madebr) [reply recipes/cppserver/all/patches/00001-cmake-clean-up.patch:9]: > What do you think?

:notes:  to my :ear:

File: recipes/cppserver/all/conanfile.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    def _build_subfolder(self):
+        return "build_subfolder"
+
+    def _configure_cmake(self):
+        if not self._cmake:
+            self._cmake = CMake(self)
+            self._cmake.definitions["CPPSERVER_MODULE"] = "OFF"
+            self._cmake.configure(build_folder=self._build_subfolder)
+        return self._cmake
+
```
Comment (uilianries) [inline recipes/cppserver/all/conanfile.py:39]:
```suggestion

    @property
    def _compilers_minimum_version(self):
        return {
            "gcc": "9",
            "Visual Studio": "15",
            "clang": "5",
            "apple-clang": "10",
        }

    def configure(self):
        if self.settings.compiler.get_safe("cppstd"):
            tools.check_min_cppstd(self, "17")
        minimum_version = self._compilers_minimum_version.get(str(self.settings.compiler), False)
        if not minimum_version:
            self.output.warn("cppserver requires C++17. Your compiler is unknown. Assuming it supports C++17.")
        elif tools.Version(self.settings.compiler.version) < minimum_version:
            raise ConanInvalidConfiguration("cppserver requires a compiler that supports at least C++17")

```

According to 5d837e3c255ff7e2e03af9e83ed0d74e3da3bc21396deb683e40a647ca8301ecR87, it requires C++17

**Review by prince-chrismc — DISMISSED** [review]

File: recipes/cppserver/all/conanfile.py [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+        os.rename(extracted_dir, self._source_subfolder)
+
+    def config_options(self):
+        if self.settings.os == "Windows":
+            del self.options.fPIC
+
+    def configure(self):
+        if self.options.shared:
+            del self.options.fPIC
+
```
Comment (uilianries) [inline recipes/cppserver/all/conanfile.py:70]:
```suggestion
```

File: recipes/cppserver/all/conanfile.py [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+    @property
+    def _compilers_minimum_version(self):
+        return {
+            "gcc": "9",
+            "Visual Studio": "15",
+            "clang": "5",
+            "apple-clang": "10",
+        }
+
+    def configure(self):
```
Comment (uilianries) [inline recipes/cppserver/all/conanfile.py:49]:
```suggestion
    def configure(self):
        if self.options.shared:
            del self.options.fPIC

```

**Review by uilianries — APPROVED** [review]

**Review by prince-chrismc — APPROVED** [review]

**Review by madebr — APPROVED** [review]
